### PR TITLE
feat: promises rewrite and drop support for node 10

### PR DESCRIPTION
**Pull request by erunion**
Late 2019 in [rdme](https://npm.im/rdme) we lost the ability to bundle in `$ref` pointers spread across multiple files, and unfortunately with the way that that library is written it's difficult to use `oas-normalize` because it's currently based on supplying callbacks. On top of that, the `swagger-parser` and `swagger2openapi` dependencies that `oas-normalize` have also been rewritten to support promises, leaving this thing in the dust.

So I've fully rewritten it to be promise-based. And added full unit test coverage in the process. And removed a handful of dependencies that it didn't *really* need. 

🚀 

## 🧰 What's being changed?

* [x] 🆕  &nbsp;Full rewrite of the library to be promise-based.
* [x] 🆕 &nbsp;Full test coverage.
* [x] 🧹  &nbsp;Pruning back unnecessary dependencies.
* [x] 💀 &nbsp;Drops support for Node 10.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,25 +15,26 @@
...
   "license": "MIT",
   "dependencies": {
-    "@apidevtools/json-schema-ref-parser": "^9.0.1",
-    "deep-extend": "^0.6.0",
-    "r2": "^2.0.1",
+    "@apidevtools/json-schema-ref-parser": "^9.0.7",
+    "node-fetch": "^2.6.1",
     "swagger-parser": "^10.0.1",
-    "swagger2openapi": "^2.11.16",
+    "swagger2openapi": "^7.0.6",
```
Comment (kanadgupta) [inline package.json:28]: omfg lol

File: package.json [hunk 2]
```diff
@@ -15,25 +15,26 @@
...
     "release": "conventional-changelog -i CHANGELOG.md -s",
     "test": "jest --coverage"
   },
   "license": "MIT",
   "dependencies": {
-    "@apidevtools/json-schema-ref-parser": "^9.0.1",
-    "deep-extend": "^0.6.0",
-    "r2": "^2.0.1",
+    "@apidevtools/json-schema-ref-parser": "^9.0.7",
+    "node-fetch": "^2.6.1",
```
Comment (kanadgupta) [inline package.json:26]: hell yeah

File: index.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                  },
+                ],
+                full: err,
+              });
+            }
+
+            return Promise.resolve({
+              errors: [
+                {
+                  message: err.message.replace(/\[object Object\]/g, 'Schema'),
```
Comment (kanadgupta) [inline index.js:107]: Nit but could `err.message` be undefined here?

**Review by kanadgupta — COMMENTED** [review]
Just did a pass and changes seem sound! I don't know much about this library and if anything, I'm just weary of all the different ways an OAS file could be invalid lol. Seeing the increase in test coverage is reassuring though!

Also one general question: should all the fixtures you created here exist in `@readme/oas-examples`? Or are you limiting that library to only valid files?

**Comment by erunion** [comment]
> Also one general question: should all the fixtures you created here exist in `@readme/oas-examples`? Or are you limiting that library to only valid files?

Yeah that library should only have valid examples.

File: package.json (hunk 1 above)
Comment (erunion) [reply package.json:28]: I've been getting pinged by Dependabot about this every month when it tries to update it. 😅

File: index.js (hunk 3 above)
Comment (erunion) [reply index.js:107]: Not from `swagger-parser`, thankfully it always throws `Error` instances with messages.

**Review by kanadgupta — APPROVED** [review]

**Pull request merged**